Core/include/Acts/Utilities/DBScan.hpp

```diff
@@ -0,0 +1,230 @@
+         bool onePointCluster = false)
+      : m_eps(epsilon),
+        m_minPoints(minPoints),
+        m_onePointCluster(onePointCluster) {}
+
+  /// @brief Cluster the input points.
+  ///
+  /// This function implement the main loop of the DBScan algorithm.
+  /// It loops over all the point and will try to start new cluster
+  /// if it finds points that have yet to be clustered.
+  ///
+  /// @param inputPoints The input points to cluster.
+  /// @param clusteredPoints Vector containing the cluster ID of each point.
+  /// @return The number of clusters (excluding noise if onePointCluster==False).
+  ///
+  int cluster(const vector_p& inputPoints, std::vector<int>& clusteredPoints) {
+    // Transform the initial vector of input point to a vector of pairs
+    // with the index of the point in the initial vector.
+    vector_pair inputPointsWithIndex;
+    for (std::size_t id = 0; id < inputPoints.size(); id++) {
+      inputPointsWithIndex.push_back(std::make_pair(inputPoints[id], id));
+    }
+    // Build the KDTree with the input points.
+    tree_t tree = tree_t(std::move(inputPointsWithIndex));
+
+    // Initialize the cluster ID to 0.
+    int clusterID = 0;
+    // By default all the points are considered as noise.
+    clusteredPoints = std::vector<int>(inputPoints.size(), -1);
+
+    // Loop over all the points
+    for (std::size_t id = 0; id < inputPoints.size(); id++) {
+      // If the point is already assigned to a cluster, skip it.
+      if (clusteredPoints[id] != -1) {
+        continue;
+      }
+      // If not we try to build a new cluster
+      std::vector<std::size_t> pointToProcess{id};
+      expendCluster(tree, inputPoints, pointToProcess, clusteredPoints,
+                    clusterID);
+      // If the cluster has been created, increment the cluster ID.
+      if (clusteredPoints[id] != -1) {
+        clusterID++;
+      }
+    }
+    if (m_onePointCluster) {
+      // If noise is present and onePointCluster is true, all the noise points
+      // are considered as individual one point clusters. Loop over all the
+      // points in the KDTree.
+      for (auto& cluster : clusteredPoints) {
+        // If the point is assigned to noise, assign it to a new cluster.
+        if (cluster == -1) {
+          cluster = clusterID;
+          clusterID++;
+        }
+      }
+    }
+    return clusterID;
+  }
+
+ private:
+  /// @brief Extend the cluster.
+  ///
+  /// This function will extend the cluster by finding all the neighbours of the
+  /// current point and assign them to the current cluster.
+  /// The KDTree is used to find the neighbours and an extra cut is used to only
+  /// keep the neighbours that are within the epsilon radius.
+  ///
+  /// @param tree The KDTree containing all the points.
+  /// @param inputPoints The vector containing the input points.
+  /// @param pointToProcess The vector containing the ids of the points that need to be
```

Comment:
```suggestion
  /// @param pointsToProcess The vector containing the ids of the points that need to be
```
Maybe change the variable `pointToProcess` to `pointsToProcess`, to show that we are dealing with a set of points.

Reply:
good idea